File: src/ios/GRTPlugin.swift
@objc(GRTPlugin) class GRTPlugin : CDVPlugin {
    override func pluginInitialize() {
        print("GRTPlugin loaded")
    }

    func createTimeSeriesClassificationData(_ command: CDVInvokedUrlCommand) {
        let instanceId = TSCDWrapper.create()
        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: instanceId),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_setInfoText', [this._id, text]);
    func TimeSeriesClassificationData_setInfoText(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_setInfoText: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = TSCDWrapper.getFromId(instanceId)
        instance?.setInfoText(command.argument(at: 2) as! String)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_getInfoText', [this._id]);
    func TimeSeriesClassificationData_getInfoText(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_getInfoText: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = TSCDWrapper.getFromId(instanceId)
        let infoText = instance?.getInfoText()

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: infoText),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_setNumDimensions', [this._id, n]);
    func TimeSeriesClassificationData_setNumDimensions(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_setNumDimensions: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let numDimensions = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_setNumDimensions: Called with wrong parameters: numDimensions"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = TSCDWrapper.getFromId(instanceId)
        instance?.setNumDimensions(numDimensions)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_getNumDimensions', [this._id]);
    func TimeSeriesClassificationData_getNumDimensions(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_getNumDimensions: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = TSCDWrapper.getFromId(instanceId)
        let numDimensions = instance?.getNumDimensions()

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: Int32(numDimensions!)),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_getNumSamples', [this._id]);
    func TimeSeriesClassificationData_getNumSamples(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_getNumSamples: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = TSCDWrapper.getFromId(instanceId)
        let numSamples = instance?.getNumSamples()

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: Int32(numSamples!)),
            callbackId:command.callbackId
        )
    }

    // window.cordova.exec(resolve, reject, pluginName, 'TimeSeriesClassificationData_addSample', [this._id, label, matrix]);
    func TimeSeriesClassificationData_addSample(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_addSample: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let label = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_addSample: Called with wrong parameters: label"),
                callbackId:command.callbackId
            )
            return
        }

        guard let matrix = command.argument(at: 2) as? NSArray else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "TimeSeriesClassificationData_addSample: Called with wrong parameters: sample"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = TSCDWrapper.getFromId(instanceId)
        let mf = MatrixFloat()

        for vector in matrix {
            let vf = VectorFloat()
            for number in vector as! NSArray {
                vf.push_back(number as! Float)
            }
            mf.push_back(vf)
        }

        instance?.addSample(label, withMatrix: mf)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    //KMeansQuantizer
    func createKMeansQuantizer(_ command: CDVInvokedUrlCommand) {
        let instanceId = KMeansQuantizerWrapper.create()
        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: instanceId),
            callbackId:command.callbackId
        )
    }

    func KMeansQuantizer_setNumClusters(_ command: CDVInvokedUrlCommand){
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_setNumClusters: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let numClusters = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_setNumClusters: Called with wrong parameters: numClusters"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = KMeansQuantizerWrapper.getFromId(instanceId)
        let result = instance?.setNumClusters(numClusters)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    func KMeansQuantizer_train(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_train: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let sampleId = command.argument(at: 1) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_train: Called with wrong parameters: sampleId"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = KMeansQuantizerWrapper.getFromId(instanceId)
        let sample = TSCDWrapper.getFromId(sampleId)
        instance?.train(sample)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    func KMeansQuantizer_quantize(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_quantize: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }

        guard let vector = command.argument(at: 1) as? NSArray else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_quantize: Called with wrong parameters: vector"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = KMeansQuantizerWrapper.getFromId(instanceId)
        let vf = VectorFloat()

        print("QUANTIZE")

        for value in vector {
            print(value)
        }

        for value in vector {
            vf.push_back(value as! Float)
        }

        instance!.quantize(vf)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    func KMeansQuantizer_getFeatureVector(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "KMeansQuantizer_getFeatureVector: called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = KMeansQuantizerWrapper.getFromId(instanceId)

        let vf = instance?.getFeatureVector()
        let array = vf!.toNSArray();

        for el in array! {
            print("ELEMENT: ")
            print(el)
        }

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: array),
            callbackId:command.callbackId
        )
    }

    // HMM
    func createHMM(_ command: CDVInvokedUrlCommand) {
        let instanceId = HMMWrapper.create()
        print("Created HMM \(instanceId!)")
        commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: instanceId),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) train:(TSCDWrapper *) sample;
    func HMM_train(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_train: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let sampleId = command.argument(at: 1) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_train: Called with wrong parameters: sampleId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let sample = TSCDWrapper.getFromId(sampleId)
        instance?.train(sample)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) predict:(MatrixFloat *) timeseries;
    func HMM_predict(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_predict: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let matrix = command.argument(at: 1) as? NSArray else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_predict: Called with wrong parameters: matrix"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = HMMWrapper.getFromId(instanceId)
        let mf = MatrixFloat()

        for vector in matrix {
            let vf = VectorFloat()
            for number in vector as! NSArray {
                vf.push_back(number as! Float)
            }
            mf.push_back(vf)
        }

        instance?.predict(mf)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK),
            callbackId:command.callbackId
        )
    }

    // - (VectorFloat *) getClassLikelihoods;
    func HMM_getClassLikelihoods(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_getClassLikelihoods: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)

        let vf = instance?.getClassLikelihoods()
        let array = vf!.toNSArray();

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: array),
            callbackId:command.callbackId
        )
    }

    // - (unsigned int) getPredictedClassLabel;
    func HMM_getPredictedClassLabel(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_getPredictedClassLabel: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let predictedClassLabel = instance?.getPredictedClassLabel()

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: Int32(predictedClassLabel!)),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setHMMType: (unsigned int) hmmType;
    func HMM_setHMMType(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let type = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: hmmType"),
                callbackId:command.callbackId
            )
            return
        }

        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setHMMType(type)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setNumStates: (unsigned int) n;
    func HMM_setNumStates(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let numStates = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: numStates"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setNumStates(numStates)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setNumSymbols: (unsigned int) n;
    func HMM_setNumSymbols(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let numSymbols = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: numSymbols"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setNumSymbols(numSymbols)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setModelType: (unsigned int) n;
    func HMM_setModelType(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let modelType = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: modelType"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setModelType(modelType)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setMinChange: (float) n;
    func HMM_setMinChange(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let minChange = command.argument(at: 1) as? Float else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: minChange"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setMinChange(minChange)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setMaxNumEpochs: (unsigned int) n;
    func HMM_setMaxNumEpochs(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let maxNumEpochs = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: maxNumEpochs"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setMaxNumEpochs(maxNumEpochs)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // - (BOOL) setNumRandomTrainingIterations: (unsigned int) n
    func HMM_setNumRandomTrainingIterations(_ command: CDVInvokedUrlCommand) {
        guard let instanceId = command.argument(at: 0) as? String else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: instanceId"),
                callbackId:command.callbackId
            )
            return
        }
        guard let numRandomTrainingIterations = command.argument(at: 1) as? UInt32 else {
            self.commandDelegate.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR,
                                messageAs: "HMM_setHMMType: Called with wrong parameters: numRandomTrainingIterations"),
                callbackId:command.callbackId
            )
            return
        }
        let instance = HMMWrapper.getFromId(instanceId)
        let result = instance?.setNumRandomTrainingIterations(numRandomTrainingIterations)

        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result!),
            callbackId:command.callbackId
        )
    }

    // DTW
    func createDTW(_ command: CDVInvokedUrlCommand) {
        let instanceId = DTWWrapper.create()
        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: instanceId),
            callbackId:command.callbackId
        )
    }
}

File: src/android/com/kano/grt/GRTPlugin.java
package com.kano.grt;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.LOG;

import java.nio.ByteBuffer;
import java.math.BigDecimal;

import org.json.JSONArray;
import org.json.JSONException;

import com.kano.grt.TimeSeriesClassificationData;

import com.kano.grt.GRTAction;
import com.kano.grt.TSCDWrapper;
import com.kano.grt.DTWWrapper;
import com.kano.grt.HMMWrapper;
import com.kano.grt.KMeansQuantizerWrapper;

@SuppressWarnings("unchecked")

public class GRTPlugin extends CordovaPlugin {

    private TSCDWrapper tscdWrapper;
    private DTWWrapper dtwWrapper;
    private HMMWrapper hmmWrapper;
    private KMeansQuantizerWrapper kMeansQuantizerWrapper;

    static {
        System.loadLibrary("GRT");
    }

    public GRTPlugin() {
        tscdWrapper = new TSCDWrapper();
        dtwWrapper = new DTWWrapper();
        hmmWrapper = new HMMWrapper();
        kMeansQuantizerWrapper = new KMeansQuantizerWrapper();
    }

    //Actions
    @Override
    public boolean execute(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        return executeTimeSeriesClassificationDataAction(action, args, callbackContext)
                || executeDTWAction(action, args, callbackContext)
                || executeHMMAction(action, args, callbackContext)
                || executeKMeansQuantizerAction(action, args, callbackContext);
    }
     
    private boolean executeTimeSeriesClassificationDataAction(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if (GRTAction.CREATE_TSCD.toString().equals(action)) {
            String id = tscdWrapper.create();
            callbackContext.success(id);
            return true;
        } else if (GRTAction.TSCD_SET_INFO_TEXT.toString().equals(action)) {
            tscdWrapper.setInfoText(args.getString(0), args.getString(1));
            callbackContext.success();
            return true;
        } else if (GRTAction.TSCD_GET_INFO_TEXT.toString().equals(action)) {
            String infoText = tscdWrapper.getInfoText(args.getString(0));
            callbackContext.success(infoText);
            return true;
        } else if (GRTAction.TSCD_SET_DATASET_NAME.toString().equals(action)) {
            tscdWrapper.setDatasetName(args.getString(0), args.getString(1));
            callbackContext.success();
            return true;
        } else if (GRTAction.TSCD_GET_DATASET_NAME.toString().equals(action)) {
            String name = tscdWrapper.getDatasetName(args.getString(0));
            callbackContext.success(name);
            return true;
        } else if (GRTAction.TSCD_SET_NUM_DIMENSIONS.toString().equals(action)) {
            boolean result = tscdWrapper.setNumDimensions(args.getString(0), args.getLong(1));
            if (result) {
                callbackContext.success();
            } else {
                callbackContext.error(0);
            }
            return true;
        } else if (GRTAction.TSCD_GET_NUM_DIMENSIONS.toString().equals(action)) {
            long n = tscdWrapper.getNumDimensions(args.getString(0));
            callbackContext.success((int) n);
            return true;
        } else if (GRTAction.TSCD_ADD_SAMPLE.toString().equals(action)) {
            String id = args.getString(0);
            int label = args.getInt(1);
            JSONArray data = args.getJSONArray(2);
            JSONArray item;
            double[][] matrix = new double[data.length()][];
            for (int i = 0; i < data.length(); i++) {
                item = data.getJSONArray(i);
                double[] row = new double[item.length()];
                for (int j = 0; j < item.length(); j++) {
                    row[j] = item.getDouble(j);
                }
                matrix[i] = row;
            }
            tscdWrapper.addSample(id, label, matrix);
            callbackContext.success();
            return true;
        } else if (GRTAction.TSCD_GET_NUM_SAMPLES.toString().equals(action)) {
            long n = tscdWrapper.getNumSamples(args.getString(0));
            callbackContext.success((int) n);
            return true;
        }
        return false;
    }

    private boolean executeDTWAction(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if (GRTAction.CREATE_DTW.toString().equals(action)) {
            String id = dtwWrapper.create();
            callbackContext.success(id);
            return true;
        } else if (GRTAction.DTW_TRAIN.toString().equals(action)) {
            TimeSeriesClassificationData t = tscdWrapper.getFromId(args.getString(1));
            boolean result = dtwWrapper.train(args.getString(0), t);
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.DTW_PREDICT.toString().equals(action)) {
            JSONArray data = args.getJSONArray(1);
            JSONArray item;
            double[][] matrix = new double[data.length()][];
            for (int i = 0; i < data.length(); i++) {
                item = data.getJSONArray(i);
                double[] row = new double[item.length()];
                for (int j = 0; j < item.length(); j++) {
                    row[j] = item.getDouble(j);
                }
                matrix[i] = row;
            }
            boolean result = dtwWrapper.predict(args.getString(0), matrix);
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.DTW_GET_MAXIMUM_LIKELIHOOD.toString().equals(action)) {
            double result = dtwWrapper.getMaximumLikelihood(args.getString(0));
            byte[] bytes = new byte[8];
            ByteBuffer.wrap(bytes).putDouble(result);
            callbackContext.success(bytes);
            return true;
        } else if (GRTAction.DTW_GET_PREDICTED_CLASS_LABEL.toString().equals(action)) {
            int result = dtwWrapper.getPredictedClassLabel(args.getString(0));
            callbackContext.success(result);
            return true;
        }
        return false;
    }

    private boolean executeKMeansQuantizerAction(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if (GRTAction.CREATE_K_MEANS_QUANTIZER.toString().equals(action)) {
            String id = kMeansQuantizerWrapper.create();
            callbackContext.success(id);
            return true;
        } else if (GRTAction.KMQ_TRAIN.toString().equals(action)) {
            TimeSeriesClassificationData t = tscdWrapper.getFromId(args.getString(1));
            boolean result = kMeansQuantizerWrapper.train(args.getString(0), t);
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.KMQ_SET_NUM_CLUSTERS.toString().equals(action)) {
            boolean result = kMeansQuantizerWrapper.setNumClusters(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.KMQ_QUANTIZE.toString().equals(action)) {
            JSONArray data = args.getJSONArray(1);
            double[] sample = new double[data.length()];
                for (int i = 0; i < data.length(); i++) {
                    sample[i] = data.getDouble(i);
                }
            int result = kMeansQuantizerWrapper.quantize(args.getString(0), sample);
            callbackContext.success(result);
            return true;
        } else if (GRTAction.KMQ_GET_FEATURE_VECTOR.toString().equals(action)) {
            double[] result = kMeansQuantizerWrapper.getFeatureVector(args.getString(0));
            JSONArray jsonResult = new JSONArray();
            for (int i = 0; i < result.length; i++) {
                jsonResult.put(result[i]);
            }
            callbackContext.success(jsonResult);
            return true;
        } else if (GRTAction.KMQ_CLEAR.toString().equals(action)) {
            boolean result = kMeansQuantizerWrapper.clear(args.getString(0));
            callbackContext.success(result ? 1 : 0);
            return true;
        }
        return false;
    }

    private boolean executeHMMAction(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if (GRTAction.CREATE_HMM.toString().equals(action)) {
            String id = hmmWrapper.create();
            callbackContext.success(id);
            return true;
        } else if (GRTAction.HMM_TRAIN.toString().equals(action)) {
            TimeSeriesClassificationData t = tscdWrapper.getFromId(args.getString(1));
            boolean result = hmmWrapper.train(args.getString(0), t);
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_PREDICT.toString().equals(action)) {
            JSONArray data = args.getJSONArray(1);
            JSONArray item;
            double[][] matrix = new double[data.length()][];
            for (int i = 0; i < data.length(); i++) {
                item = data.getJSONArray(i);
                double[] row = new double[item.length()];
                for (int j = 0; j < item.length(); j++) {
                    row[j] = item.getDouble(j);
                }
                matrix[i] = row;
            }
            boolean result = hmmWrapper.predict(args.getString(0), matrix);
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_GET_CLASS_LIKELIHOODS.toString().equals(action)) {
            double[] result = hmmWrapper.getClassLikelihoods(args.getString(0));
            JSONArray jsonResult = new JSONArray();
            for (int i = 0; i < result.length; i++) {
                LOG.w("DRFHUISJFGIUSDHFGHBDS", Double.toString(result[i]));
                jsonResult.put(result[i]);
            }
            callbackContext.success(jsonResult);
            return true;
        } else if (GRTAction.HMM_GET_PREDICTED_CLASS_LABEL.toString().equals(action)) {
            int result = hmmWrapper.getPredictedClassLabel(args.getString(0));
            callbackContext.success(result);
            return true;
        } else if (GRTAction.HMM_SET_HMM_TYPE.toString().equals(action)) {
            boolean result = hmmWrapper.setHMMType(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_NUM_STATES.toString().equals(action)) {
            boolean result = hmmWrapper.setNumStates(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_NUM_SYMBOLS.toString().equals(action)) {
            boolean result = hmmWrapper.setNumSymbols(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_MODEL_TYPE.toString().equals(action)) {
            boolean result = hmmWrapper.setModelType(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_MIN_CHANGE.toString().equals(action)) {
            boolean result = hmmWrapper.setMinChange(args.getString(0), BigDecimal.valueOf(args.getDouble(1)).floatValue());
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_MAX_NUM_EPOCHS.toString().equals(action)) {
            boolean result = hmmWrapper.setMaxNumEpochs(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        } else if (GRTAction.HMM_SET_NUM_RANDOM_TRAINING_ITERATIONS.toString().equals(action)) {
            boolean result = hmmWrapper.setNumRandomTrainingIterations(args.getString(0), args.getInt(1));
            callbackContext.success(result ? 1 : 0);
            return true;
        }
        return false;
    }
}
